### PR TITLE
Support multiple session providers

### LINKED ISSUE #596: Save session into the PostgreSQL

### PR DESCRIPTION
**Pull request by DementevNikita**
closes #596

### CODE REVIEW

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/auth/session/JooqSessionRepository.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        return Mono
+            .defer(() -> {
+                if (!session.getId().equals(session.getOriginalId())) {
+                    return deleteById(session.getOriginalId());
+                }
+
+                return Mono.empty();
+            })
+            .then(jooqReactiveOperations.mono(springSessionQuery))
+            .thenMany(Flux.defer(() -> {
```
Comment (damirabdul) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/auth/session/JooqSessionRepository.java:67]: why do you need defer here?

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/util/JooqQueryHelper.java [hunk 2]
```diff
@@ -7,6 +7,8 @@
 import lombok.RequiredArgsConstructor;
 import org.jooq.DSLContext;
 import org.jooq.Field;
+import org.jooq.InsertResultStep;
```
Comment (damirabdul) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/util/JooqQueryHelper.java:10]: please remove unused imports

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/auth/session/PostgreSQLSessionHousekeepingJobHandler.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.scheduling.annotation.Scheduled;
+import java.util.concurrent.TimeUnit;
+
+@RequiredArgsConstructor
+@Slf4j
+public class PostgreSQLSessionHousekeepingJobHandler {
+    private final PostgreSQLSessionHousekeepingJob job;
+
+    @Scheduled(fixedRate = 1, timeUnit = TimeUnit.HOURS)
+    public void deleteExpiredSessions() {
```
[the remark this answers is not in the document]
Comment (DementevNikita) [reply odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/auth/session/PostgreSQLSessionHousekeepingJobHandler.java:14]: That must be sufficient. There are three simple database operations within the job and it runs hourly.

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/auth/session/JooqSessionRepository.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+import static org.opendatadiscovery.oddplatform.model.Tables.SPRING_SESSION_ATTRIBUTES;
+
+@RequiredArgsConstructor
+@Slf4j
+public class JooqSessionRepository implements ReactiveSessionRepository<MapSession> {
+    private final JooqReactiveOperations jooqReactiveOperations;
+
+    @Override
+    public Mono<MapSession> createSession() {
+        return Mono.defer(() -> Mono.just(new MapSession()));
```
[the remark this answers is not in the document]
Comment (DementevNikita) [reply odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/auth/session/JooqSessionRepository.java:39]: Spring Session lib calls this method in the switchIfEmpty.
```
@Override
public Mono<WebSession> getSession(ServerWebExchange exchange) {
     return Mono.defer(() -> retrieveSession(exchange)
		.switchIfEmpty(createWebSession())
		.doOnNext(session -> exchange.getResponse().beforeCommit(() -> save(exchange, session))));
}

```

**Review by damirabdul — APPROVED** [review]
LGTM

**Pull request merged**